internal/juju/secrets.go

```
@@ -0,0 +1,199 @@
+	if err != nil {
+		return CreateSecretOutput{}, typedError(err)
+	}
+	return CreateSecretOutput{
+		SecretId: secretId,
+	}, nil
+}
+
+// ReadSecret reads a secret.
+func (c *secretsClient) ReadSecret(input *ReadSecretInput) (ReadSecretOutput, error) {
+	conn, err := c.GetConnection(&input.ModelName)
+	if err != nil {
+		return ReadSecretOutput{}, err
+	}
+	defer func() { _ = conn.Close() }()
+
+	secretAPIClient := c.getSecretAPIClient(conn)
+
+	var secretURI *coresecrets.URI
+	if input.SecretId != "" {
+		secretURI, err = coresecrets.ParseURI(input.SecretId)
+		if err != nil {
+			return ReadSecretOutput{}, err
+		}
+	} else {
+		secretURI = nil
+	}
+	secretFilter := coresecrets.Filter{
+		URI:      secretURI,
+		Label:    input.Name,
+		Revision: input.Revision,
+	}
+
+	results, err := secretAPIClient.ListSecrets(true, secretFilter)
+	if err != nil {
+		return ReadSecretOutput{}, typedError(err)
+	}
+	if len(results) < 1 {
+		return ReadSecretOutput{}, &secretNotFoundError{secretId: input.SecretId}
+	}
+	if results[0].Error != "" {
+		return ReadSecretOutput{}, errors.New(results[0].Error)
+	}
+
+	value, err := results[0].Value.Values()
+	if err != nil {
+		return ReadSecretOutput{}, err
+	}
+	return ReadSecretOutput{
+		Value: value,
+	}, nil
+}
+
+// UpdateSecret updates a secret.
+func (c *secretsClient) UpdateSecret(input *UpdateSecretInput) error {
+	conn, err := c.GetConnection(&input.ModelName)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+
+	secretAPIClient := c.getSecretAPIClient(conn)
+
+	// Specify by ID or Name
+	if input.SecretId == "" && input.Name == nil {
+		return errors.New("must specify either secret ID or name")
+	}
+	if input.SecretId != "" && input.Name != nil {
+		return errors.New("must specify either secret ID or name but not both")
+	}
+
+	// Specify by name
+	if input.SecretId == "" {
+		err = secretAPIClient.UpdateSecret(nil, *input.Name, input.AutoPrune, *input.NewName, *input.Info, *input.Value)
```

Could these de-references result in panics if the point is nil? The points arn't checked anywhere.
Also err is not checked.

---

thx, fixed